core/src/main/scala/kafka/metrics/clientmetrics/ClientMetricsConfig.scala

```diff
@@ -0,0 +1,136 @@
+package kafka.metrics.clientmetrics
+
+import kafka.metrics.clientmetrics.ClientMetricsCacheOperation.{CM_SUBSCRIPTION_ADDED, CM_SUBSCRIPTION_DELETED, CM_SUBSCRIPTION_UPDATED}
+import kafka.metrics.clientmetrics.ClientMetricsConfig.ClientMetrics.{AllMetricsFlag, ClientMatchPattern, DeleteSubscription, PushIntervalMs, SubscriptionMetrics, configDef}
+import org.apache.kafka.common.config.ConfigDef
+import org.apache.kafka.common.config.ConfigDef.Importance.MEDIUM
+import org.apache.kafka.common.config.ConfigDef.Type.{BOOLEAN, INT, LIST}
+import org.apache.kafka.common.errors.{InvalidConfigurationException, InvalidRequestException}
+
+import java.util
+import java.util.Properties
+import java.util.concurrent.ConcurrentHashMap
+import scala.collection.mutable
+import scala.collection.mutable.ListBuffer
+
+/**
+ * Client metric configuration related parameters and the supporting methods like validation and update methods
+ * are defined in this class.
+ * For more information please look at kip-714:
+ * https://cwiki.apache.org/confluence/display/KAFKA/KIP-714%3A+Client+metrics+and+observability
+ */
+object ClientMetricsConfig {
+
+  class SubscriptionGroup(subscriptionGroup: String,
+                          subscribedMetrics: List[String],
+                          var matchingPatternsList: List[String],
+                          pushIntervalMs: Int,
+                          allMetricsSubscribed: Boolean = false) {
+    def getId = subscriptionGroup
+    def getPushIntervalMs = pushIntervalMs
+    val clientMatchingPatterns = parseClientMatchingPatterns(matchingPatternsList)
+    def getClientMatchingPatterns = clientMatchingPatterns
+    def getSubscribedMetrics = subscribedMetrics
+    def getAllMetricsSubscribed = allMetricsSubscribed
+
+    private def parseClientMatchingPatterns(patterns: List[String]) : Map[String, String] = {
+      val patternsMap = mutable.Map[String, String]()
+      patterns.foreach(x =>  {
+        val nameValuePair = x.split("=").map(x => x.trim)
+        if (nameValuePair.size != 2) {
+          throw new InvalidConfigurationException("Illegal client matching pattern: " + x)
+        }
+        patternsMap += (nameValuePair(0) -> nameValuePair(1))
+      })
+      patternsMap.toMap
+    }
+  }
+
+  object ClientMetrics {
+    // Properties
```

Comment:
Is this the configuration for the subscription information? Is this statically defined in a properties file?

I thought subscriptions and their settings were dynamically modifiable at multiple "levels" (broker-wide, per-client, per-client type, etc.).

Reply:
For now the expectation is kafka_metrics_config.sh/AdminClient would send these parameters along with the RPC messages (DescribeConfigRequest/AlterConfigRequest). However when we adopt this feature in CDC we may need to use some kind of static configuration.  The main reason for using the ConfigDef class to use its parsing and validating methods.

Reply:
Makes sense. Thanks!